src/transbank.c

```diff
@@ -0,0 +1,242 @@
+    portList = malloc((separators) + chars * sizeof(char*));
+    if (portList != NULL ){
+      for (int i = 0; i < separators; i++) {
+        strcat(portList, sp_get_port_name(ports[i]));
+        if (i < separators -1)
+        {
+          strcat(portList, "|");
+        }
+      }
+    }
+  }else {
+    char* error = "No serial devices detected\n";
+    portList = malloc(strlen(error) * sizeof(char*));
+    strcpy(portList, error);
+  }
+  sp_free_port_list(ports);
+  return portList;
+}
+
+char* get_configured_port_name(){
+  if (port != NULL)
+    return sp_get_port_name(port);
+  else
+    return "No port configured";
+}
+
+int select_port(char* portName) {
+  int retval = sp_get_port_by_name(portName, &port);
+  return retval;
+}
+
+int open_configured_port(){
+  return sp_open(port, SP_MODE_WRITE | SP_MODE_READ);
+}
+
+int configure_port(int baud_rate) {
+  int retval = 0;
+  struct sp_port_config *config;
+
+  retval += sp_new_config(&config);
+  retval += sp_set_config_baudrate(config, baud_rate);
+  retval += sp_set_config_bits(config, BITS);
+  retval += sp_set_config_parity(config, PARITY);
+  retval += sp_set_config_stopbits(config, STOP_BITS);
+
+  retval += sp_set_config(port, config);
+  retval += sp_flush(port, SP_BUF_BOTH);
+  sp_free_config(config);
+
+  return retval;
+}
+
+int write_message(Message message){
+  int retval = sp_blocking_write(port, message.payload, message.payloadSize, DEFAULT_TIMEOUT);
+  if (retval == message.payloadSize && sp_drain(port))
+  {
+    retval = TBK_OK;
+  }
+  else
+    retval -= message.payloadSize;
+  sp_flush(port, SP_BUF_OUTPUT);
+  return retval;
+}
+
+int reply_ack(){
+  char buf[] = {ACK};
+  return sp_blocking_write(port, buf, 1, DEFAULT_TIMEOUT);
+}
+
+int read_bytes(char* buf, Message message){
+  memset(buf, 0, message.responseSize);
+  int retval = sp_blocking_read_next(port, buf, message.responseSize, DEFAULT_TIMEOUT);
+  if (retval == message.responseSize)
+    retval = TBK_OK;
+  else{
+    retval -= message.responseSize;
+  }
+  sp_flush(port, SP_BUF_INPUT);
+  return retval;
+}
+
+int read_ack(){
+  char buf[1];
+  int retval = sp_blocking_read_next(port, buf, 1, DEFAULT_TIMEOUT);
+  if (retval == 1 && buf[0] == ACK)
+    return TBK_OK;
+  else
+    return TBK_NOK;
+}
+
+int get_totals(){
+  int tries = 0;
+  do
+  {
+    int retval = write_message(GET_TOTALS);
+    if (retval == TBK_OK){
+      char* returnBuf;
+      retval = read_bytes(returnBuf, GET_TOTALS);
+      if (retval == TBK_OK && returnBuf[0] == ACK)
+        return TBK_OK;
+    }
+    tries++;
+  } while (tries < GET_TOTALS.retries);
+  return TBK_NOK;
+}
+
+int load_keys(){
+  int tries = 0;
+  do
+  {
+    int retval = write_message(LOAD_KEYS);
+    if (retval == TBK_OK){
+      char* returnBuf;
+      retval = read_ack();
+      if (retval == TBK_OK && returnBuf[0] == ACK){
+        retval = read_bytes(returnBuf, LOAD_KEYS);
+        if (retval == TBK_OK)
+          for(int i=0; i < sizeof(returnBuf); i++)
```

Comment:
This line code is fine, but beware that avoiding using curly braces could lead to a misunderstood when modifying the code in the future.

As a curiosity, Apple's SSL bug was avoidable using an always-braces rule.

Take it or leave it :)